### PR TITLE
Weekly/Monthly Habit reset counters resetting early - fixes #8570

### LINKED ISSUE #8570: weekly (and monthly?) Habit counters are resetting too often
The new Habit counters added in PR https://github.com/HabitRPG/habitica/pull/8198 have a setting for resetting either each day or week or month. Currently, the weekly counters are always resetting each day, regardless of what setting is chosen. There is some evidence that monthly counters might sometimes also be resetting too soon but that's not certain (see one of the reports below).

I tested overnight by having a Habit set to each of the reset frequencies, with clicks recorded in each of their counters. The daily and weekly Habits were reset this morning and the monthly one was not. The weekly one should not have been reset. I've copied other people's reports from the Report a Bug guild at the bottom of this post.

Recent releases after the counters went live have made changes related to dates and cron, so it's not impossible that this feature broke after going live.
https://github.com/HabitRPG/habitica/releases/tag/v3.79.2  https://github.com/HabitRPG/habitica/releases/tag/v3.79.1
One of the reports from the Bug guild hints at that: "Mine (weekly) were saving fine til Sunday reset then I noticed they reset again this AM". Those two releases happened around Sunday, depending on time zone.

Worryingly though, the automatic tests aren't picking up the bug, even though I can see that there are tests for this behaviour. 

I'm marking this as high priority because the counters are a solution to several common requests from users for better tracking of Habits and it would be good to get the resets working again soon.

@shalott If you have any ideas about what might be causing the resets to not work, we'd love to hear from you!

From the Report a Bug guild:

apolwapol (26677250-949e-4f84-b22b-3815f7ed5d5d):
Hi, I set the habit counter to reset weekly, but it still resets daily.
-- 18 hours ago 

Yayef (23d6c69b-b1f6-4f02-940c-c5417ef3e1a1):
Hello. Same issue as @apolwapol : all my habits counters (weekly or monthly) reset daily.
-- 15 hours ago

M. Hostile (bea18590-b955-4c2b-9005-ec714fbcdbaa):
Hi! I like the new feature of habit counters but there seems to be a bug. My weekly counters reset every day. Monthly counters work fine. Thanks for your work.
-- 12 hours ago   

joedragons (66902097-c79b-496f-acd9-b78cc5867b28):
Re habit counters: Mine (weekly) were saving fine til Sunday reset then I noticed they reset again this AM, in case that helps. Could be the first one works then after that gets off. Or daylight savings time (if you respect that=)). Or something changed around then? I just assumed they reset Sunday, maybe that's not true=) I turned them weekly the first day they were announced if that helps. Appreciate the investigation.
-- 9 hours ago

### PR DESCRIPTION
**Pull request by joe-salomon2**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes https://github.com/HabitRPG/habitica/issues/8570

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

Previously, weekly/monthly habit reset counters were using server time to calculate "Monday" or "First of the month" to reset. This caused counters to reset too often if a user's timezone was already the next day, but the server was still on a Monday or "First day". Also, counters would reset early if the user was behind UTC.

For habit reset logic, changed day check calculation to use user’s timezone instead of server time.
Added unit tests to check following cases:
- Weekly habit reset: Server tz is Sunday, User tz is Monday
- Weekly habit reset: Server tz is Monday, User tz is Sunday
- Monthly habit reset: Server tz is 1st of month, User tz is 2nd of month
- Monthly habit reset: Server tz is end of prev month, User tz is 1st of month

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: fc28e9dc-9c84-4ee3-be6e-b4dffa54f558

### CODE REVIEW

File: website/server/libs/cron.js [hunk 1]
```diff
@@ -329,11 +329,11 @@ export function cron (options = {}) {
     if (resetWeekly === true && resetMonthly === true) {
       break;
     }
-    let thatDay = moment(now).subtract({days: i}).toDate();
-    if (thatDay.getDay() === 1) {
+    let thatDay = moment(now).utcOffset(user.preferences.timezoneOffset).subtract({days: i});
```
Comment (TheHollidayInn) [inline website/server/libs/cron.js]: I'm pretty sure this needs to be zone because we haven't been upgrading to utcOffset.

Comment (TheHollidayInn) [reply website/server/libs/cron.js]: See here: https://github.com/HabitRPG/habitica/blob/develop/website/common/script/cron.js#L106

Comment (joe-salomon2) [reply website/server/libs/cron.js]: Got it, will fix for consistency

**Comment by TheHollidayInn** [comment]
This does fix the weekly reset part for me locally. Although, I'm not quite convinced this will fix CDS issues.

**Comment by joe-salomon2** [comment]
Do you suggest I use startOfDayWithCDSTime as well? Referenced in that gist above.

**Comment by TheHollidayInn** [comment]
I was just showing how we use zone instead. Also, if you have a way to check manually with the CDS, to show it fail and succeed, that would be helpful in verifying.

**Comment by joe-salomon2** [comment]
Gotcha, I can fix the usage of zone.
For CDS, do you mean a test that uses something like this variable: https://github.com/HabitRPG/habitica/blob/develop/website/common/script/cron.js#L101, which uses a date without the user's time portion associated with it?

**Comment by TheHollidayInn** [comment]
I think your tests are great. I mean, it would be great to document a way to test this in the browser without a unit/integration test.

**Comment by joe-salomon2** [comment]
Interestingly enough, the unit tests work, but the wrong behavior is still there on the site. Let me investigate some more.

**Comment by Alys** [comment]
@joe-salomon When we merge PRs, they don't go live immediately so your fix isn't on the site yet.  Have a look at ["Acceptance of Your PR" in Using Your Local Install to Modify Habitica's Website and API](http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API#Acceptance_of_Your_PR) and ask here if you any questions (we would then improve that documentation).

**Comment by joe-salomon2** [comment]
Ah no I meant on my locally running version.

**Comment by joe-salomon2** [comment]
Fixed the issues

Steps to test on the Habitica site:
Test Case 1: Default day start of 12am
1. Set Custom Day Start to 12am
2. Set computer clock date to any Saturday after user's lastCron value in DB.
3. Create Weekly Habit
4. Increase Habit multiple times
5. Set computer clock to following Sunday
6. Refresh Habitica page
_Note Habit counter has NOT reset_
7. Set computer clock to following Monday
8. Refresh Habitica page
_Note Habit counter HAS reset_
9. Increase Habit multiple times
10. Set computer clock to following Tuesday
_Note Habit counter has NOT reset_

Test Case 2: Custom Day Start of 3am
1. Set Custom Day Start to 3am
2. Set computer clock date to any Saturday after 3am, after user's lastCron value in DB.
3. Create Weekly Habit
4. Increase Habit multiple times
5. Set computer clock to following Sunday after 3am
6. Refresh Habitica page
_Note Habit counter has NOT reset_
7. Set computer clock to following Monday at 2am (before cron)
8. Refresh Habitica page
_Note Habit counter has NOT reset_
9. Set computer clock to following Monday at 4am (after cron)
10. Refresh Habitica page
_Note Habit counter HAS reset_
11. Increase Habit multiple times
12. Set computer clock to following Tuesday after 3am
_Note Habit counter has NOT reset_

**Comment by joe-salomon2** [comment]
Hey guys, any issues with this one still?

**Comment by TheHollidayInn** [comment]
Hey, @joe-salomon! We review every other week. I'll switch this over to ready for review. Sorry about that.

**Comment by SabreCat** [comment]
This is fantastic, @joe-salomon. Sorry it took so long for me to complete review on it! Added this as credit toward Blacksmith tier 3 for you.

**Pull request merged**